tiledb/sm/filesystem/s3.cc

```diff
@@ -947,40 +976,53 @@ Status S3::init_client() const {
   }
 #endif
 
+  if (!aws_access_key_id.empty() && !aws_secret_access_key.empty() && !aws_role_arn.empty()){
+    return Status::S3Error("Authentication credentials can exclusively be either long lived access credentials"
+        "or temporal ones assumed by the role");
+  }
+
   // If the user set config variables for AWS keys use them.
   if (!aws_access_key_id.empty() && !aws_secret_access_key.empty()) {
     Aws::String access_key_id(aws_access_key_id.c_str());
     Aws::String secret_access_key(aws_secret_access_key.c_str());
-    client_creds_ = std::unique_ptr<Aws::Auth::AWSCredentials>(
-        new Aws::Auth::AWSCredentials(access_key_id, secret_access_key));
-
-    // If the user has set a session token (for AWS Security Token Service)
-    // then use it:
-    //     - https://docs.aws.amazon.com/STS/latest/APIReference/Welcome.html
-    // For testing run: `aws sts get-session-token --duration-seconds 900`. See:
-    //     -
-    //     https://docs.aws.amazon.com/cli/latest/reference/sts/get-session-token.html
-    if (!aws_session_token.empty()) {
-      Aws::String session_token(aws_session_token.c_str());
-      client_creds_->SetSessionToken(session_token);
-    }
+    Aws::String session_token(
+        !aws_session_token.empty() ? aws_session_token.c_str() : "");
+    credentials_provider_ =
+        std::make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(
+            access_key_id, secret_access_key, session_token);
   }
 
-  if (client_creds_ == nullptr) {
-    client_ = Aws::MakeShared<Aws::S3::S3Client>(
-        constants::s3_allocation_tag.c_str(),
-        *client_config_,
-        Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Never,
-        use_virtual_addressing_);
-  } else {
+  // If AWS Role ARN provided instead of access_key and secret_key,
+  // temporary credentials will be fetched by assuming this role.
+  else if (!aws_role_arn.empty()) {
+    Aws::String role_arn(aws_role_arn.c_str());
+    Aws::String external_id(
+        !aws_external_id.empty() ? aws_external_id.c_str() : "");
+    int load_frequency(
+        !aws_load_frequency.empty() ? std::stoi(aws_load_frequency) : 999);
+    Aws::String session_token(
+        !aws_session_token.empty() ? aws_session_token.c_str() : "");
+    credentials_provider_ =
+        std::make_shared<Aws::Auth::STSAssumeRoleCredentialsProvider>(
+            role_arn, session_token, external_id, load_frequency, nullptr);
+  }
+  else{
```

Comment:
For readability, let's remove this `else` and instead do:
```
if (credentials_provider_ == nullptr) {
  client_ = ...
} else {
  client_ = ...
}
```

Reply:
Do you think that since I remove this else it would be better to change also the above statements, so it doesn't terminate with `else if` and just provide a more generic error message in case the configuration is not correct?